Add KH instability analytic data for GRMHD

nilsdeppe opened the pull request:
## Proposed changes

Adds the KH instability initial data and executable.

### Upgrade instructions

<!--
If this PR makes changes that other people should be aware of when upgrading
their code, describe what they should do between the two UPGRADE INSTRUCTIONS
lines below.
-->
<!-- UPGRADE INSTRUCTIONS -->

<!-- UPGRADE INSTRUCTIONS -->

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

nilsdeppe committed:
Add GRMHD Kelvin-Helmholtz instability analytic data

nilsdeppe committed:
Add GRMHD KH instability executable

nilsdeppe commented:
Marking as priority since we need this for the referee report

yoonso0-0 commented on src/PointwiseFunctions/AnalyticData/GrMhd/KhInstability.hpp:
```diff
@@ -0,0 +1,275 @@
+  };
+
+  /// The uniform magnetic field
+  struct MagneticField {
```
Please mention magnetic field at the analytic data description in docs.

yoonso0-0 commented on src/PointwiseFunctions/AnalyticData/GrMhd/KhInstability.hpp:
```diff
@@ -0,0 +1,275 @@
+                             PerturbAmplitude, PerturbWidth, MagneticField>;
+
+  static constexpr Options::String help = {
+      "Initial data to simulate the KH instability."};
```
Add "magnetized"?

yoonso0-0 commented on src/PointwiseFunctions/AnalyticData/GrMhd/KhInstability.cpp:
```diff
@@ -0,0 +1,226 @@
+  const size_t n_pts = get_size(get<0>(x));
+  for (size_t s = 0; s < n_pts; ++s) {
+    get_element(get(result), s) =
+        abs(get_element(get<2>(x), s) - strip_bimedian_height_) <
```
Need to use `get<1>` to retrieve y coordinates

yoonso0-0 commented on src/PointwiseFunctions/AnalyticData/GrMhd/KhInstability.cpp:
```diff
@@ -0,0 +1,226 @@
+  const size_t n_pts = get_size(get<0>(x));
+  for (size_t s = 0; s < n_pts; ++s) {
+    get_element(get<0>(result), s) =
+        abs(get_element(get<2>(x), s) - strip_bimedian_height_) <
```
Same as above

yoonso0-0 commented on src/PointwiseFunctions/AnalyticData/GrMhd/KhInstability.cpp:
```diff
@@ -0,0 +1,226 @@
+      strip_bimedian_height_ - strip_half_thickness_;
+  const double strip_upper_bound =
+      strip_bimedian_height_ + strip_half_thickness_;
+  get<2>(result) =
```
Same as above (all `get<2>`s from line 131 to 134)

yoonso0-0 commented on src/PointwiseFunctions/AnalyticData/GrMhd/KhInstability.cpp:
```diff
@@ -0,0 +1,226 @@
+     hydro::Tags::Pressure, hydro::Tags::DivergenceCleaningField,
+     hydro::Tags::SpecificEnthalpy, hydro::Tags::LorentzFactor))
+
+#define INSTANTIATE_VELOCITY(_, data)                                         \
```
Maybe change macro name to `INSTANTIATE_VECTORS`?

yoonso0-0 commented on src/PointwiseFunctions/AnalyticData/GrMhd/KhInstability.cpp:
```diff
@@ -0,0 +1,226 @@
+GENERATE_INSTANTIATIONS(INSTANTIATE_VELOCITY, (double, DataVector),
+                        (hydro::Tags::SpatialVelocity,
+                         hydro::Tags::MagneticField))
+}  // namespace grmhd::AnalyticData
```
Please add `#undef`s for macro definitions

yoonso0-0 commented on tests/Unit/PointwiseFunctions/AnalyticData/GrMhd/KhInstability.py:
```diff
@@ -0,0 +1,89 @@
+                      perturbation_amplitude, perturbation_width,
+                      magnetic_field):
+    return strip_density if np.absolute(
+        x[x.size - 1] -
```
`x.size-1` -> `1`

yoonso0-0 commented on tests/Unit/PointwiseFunctions/AnalyticData/GrMhd/KhInstability.py:
```diff
@@ -0,0 +1,89 @@
+        strip_thickness) else background_velocity
+    strip_lower_bound = strip_bimedian_height - 0.5 * strip_thickness
+    strip_upper_bound = strip_bimedian_height + 0.5 * strip_thickness
+    result[dim - 1] = (
```
All `dim - 1` s  -> `1`

yoonso0-0 commented on tests/Unit/PointwiseFunctions/AnalyticData/GrMhd/KhInstability.py:
```diff
@@ -0,0 +1,89 @@
+    dim = x.size
+    result = np.zeros(dim)
+    result[0] = strip_velocity if (
+        np.absolute(x[x.size - 1] - strip_bimedian_height) < 0.5 *
```
`x.size - 1` -> `1`

yoonso0-0 commented on tests/Unit/PointwiseFunctions/AnalyticData/GrMhd/KhInstability.py:
```diff
@@ -0,0 +1,89 @@
+               ((x[dim - 1] - strip_lower_bound) / perturbation_width)**2) +
+        np.exp(-0.5 *
+               ((x[dim - 1] - strip_upper_bound) / perturbation_width)**2))
+    result[dim - 1] *= perturbation_amplitude * np.sin(4 * np.pi * x[0])
```
Here as well

yoonso0-0 commented on tests/Unit/PointwiseFunctions/AnalyticData/GrMhd/KhInstability.py:
```diff
@@ -0,0 +1,89 @@
+                             perturbation_amplitude, perturbation_width,
+                             magnetic_field):
+    return pressure / (adiabatic_index - 1.0) / strip_density if np.absolute(
+        x[x.size - 1] -
```
`x.size - 1` -> `1`

yoonso0-0 commented on tests/Unit/PointwiseFunctions/AnalyticData/GrMhd/Test_KhInstability.cpp:
```diff
@@ -0,0 +1,96 @@
+      "PointwiseFunctions/AnalyticData/GrMhd"};
+
+  test(std::numeric_limits<double>::signaling_NaN());
+  // test(DataVector(5));
```
Q. Any reason why test for `DataVector` was skipped?

nilsdeppe replied:
Left over from debugging. Thanks for catching :)

yoonso0-0 reviewed (changes requested):
Please squash.

nilsdeppe commented:
@yoonso0-0 I've squashed in the fixes to your comments. I also changed the direction the strip is in (y instead of z) because this meant I could easily re-use ParaView scripts I have from other 2d simulations of GRMHD. I think that'll generally be useful

yoonso0-0 commented:
Great :)
